cmd/ovirt-populator/ovirt-populator.go

```diff
@@ -32,154 +33,163 @@ type TransferProgress struct {
+		klog.Fatalf("Failed to write to %s: %v", filename, err)
 	}
+}
 
-	args := createCommandArguments(&engineConfig, diskID, volPath)
+func executePopulationProcess(config *engineConfig, diskID, volPath, ownerUID string, pvcSize int64) {
+	args := createCommandArguments(config, diskID, volPath)
 	cmd := exec.Command("/usr/local/bin/ovirt-img", args...)
 	r, _ := cmd.StdoutPipe()
 	cmd.Stderr = cmd.Stdout
 	done := make(chan struct{})
 	scanner := bufio.NewScanner(r)
 	klog.Info(fmt.Sprintf("Running command: %s", cmd.String()))
 
-	go func() {
-		for scanner.Scan() {
-			progressOutput := TransferProgress{}
-			text := scanner.Text()
-			klog.Info(text)
-			err = json.Unmarshal([]byte(text), &progressOutput)
-			if err != nil {
-				var syntaxError *json.SyntaxError
-				if !errors.As(err, &syntaxError) {
-					klog.Error(err)
-				}
-			}
-
-			progressGague.WithLabelValues(diskID).Set(float64(progressOutput.Transferred))
-		}
-
-		done <- struct{}{}
-	}()
+	go monitorProgress(scanner, ownerUID, pvcSize, done)
 
-	err = cmd.Start()
-	if err != nil {
+	if err := cmd.Start(); err != nil {
 		klog.Fatal(err)
 	}
 
 	<-done
-	err = cmd.Wait()
-	if err != nil {
+	if err := cmd.Wait(); err != nil {
 		klog.Fatal(err)
 	}
 }
 
-func createCommandArguments(config *engineConfig, diskID, volPath string) []string {
-	if config.insecure {
-		return []string{
-			"download-disk",
-			"--output", "json",
-			"--engine-url=" + config.URL,
-			"--username=" + config.username,
-			"--password-file=/tmp/ovirt.pass",
-			"--insecure",
-			"-f", "raw",
-			diskID,
-			volPath,
-		}
+func monitorProgress(scanner *bufio.Scanner, ownerUID string, pvcSize int64, done chan struct{}) {
+	progress := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "ovirt_progress",
+			Help: "Progress of volume population",
+		},
+		[]string{"ownerUID"},
+	)
+	if err := prometheus.Register(progress); err != nil {
+		klog.Error("Prometheus progress gauge not registered:", err)
+		return
 	} else {
-		// for secure connection use the ca cert
-		cert, err := os.Create("/tmp/ca.pem")
-		if err != nil {
-			klog.Fatalf("Failed to create ca.pem %v", err)
-		}
+		klog.Info("Prometheus progress gauge registered.")
+	}
 
-		defer cert.Close()
-		_, err = cert.Write([]byte(config.cacert))
-		if err != nil {
-			klog.Fatalf("Failed to write CA to file: %v", err)
+	var currentProgress float64
+	total := pvcSize
+	metric := &dto.Metric{}
+
+	for scanner.Scan() {
+		progressOutput := TransferProgress{}
+		text := scanner.Text()
+		klog.Info(text)
+		if err := json.Unmarshal([]byte(text), &progressOutput); err != nil {
+			var syntaxError *json.SyntaxError
+			if !errors.As(err, &syntaxError) {
+				klog.Error(err)
+			}
 		}
-
-		return []string{
-			"download-disk",
-			"--output", "json",
-			"--engine-url=" + config.URL,
-			"--username=" + config.username,
-			"--password-file=/tmp/ovirt.pass",
-			"--cafile=" + "/tmp/ca.pem",
-			"-f", "raw",
-			diskID,
-			volPath,
+		if total > 0 {
```

Comment:
that's related to my previous comment - does it even make sense that total == 0?

Reply:
no, just a precaution to avoid dividing by zero